src/contracts/AbstractARM.sol

```diff
@@ -0,0 +1,693 @@
+
+        require(request.claimTimestamp <= block.timestamp, "Claim delay not met");
+        // If there isn't enough reserved liquidity in the queue to claim
+        require(request.queued <= withdrawsClaimable, "Queue pending liquidity");
+        require(request.withdrawer == msg.sender, "Not requester");
+        require(request.claimed == false, "Already claimed");
+
+        // Store the request as claimed
+        withdrawalRequests[requestId].claimed = true;
+        // Store the updated claimed amount
+        withdrawsClaimed += request.assets;
+
+        assets = request.assets;
+
+        emit RedeemClaimed(msg.sender, requestId, assets);
+
+        // transfer the liquidity asset to the withdrawer
+        IERC20(liquidityAsset).transfer(msg.sender, assets);
+    }
+
+    /// @dev Updates the claimable amount in the ARM's withdrawal queue.
+    /// That's the amount that is used to check if a request can be claimed or not.
+    function _updateWithdrawalQueueLiquidity() internal {
+        // Load the claimable amount from storage into memory
+        uint256 withdrawsClaimableMem = withdrawsClaimable;
+
+        // Check if the claimable amount is less than the queued amount
+        uint256 queueShortfall = withdrawsQueued - withdrawsClaimableMem;
+
+        // No need to do anything is the withdrawal queue is fully funded
+        if (queueShortfall == 0) {
+            return;
+        }
+
+        uint256 liquidityBalance = IERC20(liquidityAsset).balanceOf(address(this));
+
+        // Of the claimable withdrawal requests, how much is unclaimed?
+        // That is, the amount of the liquidity assets that is currently allocated for the withdrawal queue
+        uint256 allocatedLiquidity = withdrawsClaimableMem - withdrawsClaimed;
+
+        // If there is no unallocated liquidity assets then there is nothing to add to the queue
+        if (liquidityBalance <= allocatedLiquidity) {
+            return;
+        }
+
+        uint256 unallocatedLiquidity = liquidityBalance - allocatedLiquidity;
+
+        // the new claimable amount is the smaller of the queue shortfall or unallocated weth
+        uint256 addedClaimable = queueShortfall < unallocatedLiquidity ? queueShortfall : unallocatedLiquidity;
+
+        // Store the new claimable amount back to storage
+        withdrawsClaimable = SafeCast.toUint128(withdrawsClaimableMem + addedClaimable);
+    }
+
+    /// @dev Calculate how much of the liquidity asset in the ARM is not reserved for the withdrawal queue.
+    // That is, it is available to be swapped.
+    function _liquidityAvailable() internal view returns (uint256) {
+        // The amount of WETH that is still to be claimed in the withdrawal queue
+        uint256 outstandingWithdrawals = withdrawsQueued - withdrawsClaimed;
+
+        // The amount of the liquidity asset is in the ARM
+        uint256 liquidityBalance = IERC20(liquidityAsset).balanceOf(address(this));
```

Comment:
⛽️ Since this function is used in the hot path of trades, we should add an early return on `outstandingWithdrawals == 0` above this line. This way if everyone has what they need, we remove the need for this external call.

Reply:
`_liquidityAvailable` returns the available WETH in the ARM. We need to do the external call to WETH to get the ARM's balance.

We could return `type(uint256).max` if `outstandingWithdrawals == 0` and let the WETH transfer fail if not enough WETH balance in the ARM.

Reply:
Actually, there's a bug in here. `_liquidityAvailable` also needs to consider WETH reserved for the accrued fees stored in `feesAccrued`.  That's other read of a storage slot unfortunately :-(

Reply:
I take that back. I don't think we need to make any liquidity guarantees to the fee collector, which is the ARM buyback contract. We can swap out WETH that is meant for the fee collector. The buy backs will just have to wait until the WETH is swapped back or claimed from a Lido withdrawal.

I've added
```Solidity
// Save gas on an external call if there are no outstanding withdrawals
if (outstandingWithdrawals == 0) return type(uint256).max;
```

Reply:
@clement-ux this scenario would be a good one to test.

Scenario: Trader swaps stETH for WETH
  Given ARM WETH balance > WETH swap out amount and outstanding withdrawals
    And ARM WETH balance < WETH swap out amount, outstanding withdrawals and accrued fees
  Then trader receives the WETH
    And the ARM receives the stETH

Reply:
@DanielVF @clement-ux WETH doesn't include an error message when it reverts on a transfer with not enough balance.

```Solidity
require(balanceOf[src] >= wad);
```

So we have this scenario

Scenario: Trader swaps stETH for WETH
  Given ARM WETH balance < WETH swap out amount
    And outstanding withdrawals == 0
  Then reverts with no error message

![Screenshot 2024-10-03 at 11 15 50 AM](https://github.com/user-attachments/assets/4e8c2eeb-26e5-435a-9eb9-1919635bcf4b)

It's ugly from an error handling perspective but worth the tradeoff of saving gas on swaps of stETH for WETH.

Reply:
> @clement-ux this scenario would be a good one to test.
> 
> Scenario: Trader swaps stETH for WETH Given ARM WETH balance > WETH swap out amount and outstanding withdrawals And ARM WETH balance < WETH swap out amount, outstanding withdrawals and accrued fees Then trader receives the WETH And the ARM receives the stETH

Added in this test: 
https://github.com/OriginProtocol/arm-oeth/blob/2bc43a1c9b82136231ba0f27641bbaa866597ad1/test/fork/LidoFixedPriceMultiLpARM/SwapExactTokensForTokens.t.sol#L298-L356